dev/DynamicDependency/PackageGraph.cpp

```diff
@@ -0,0 +1,396 @@
+
+    // Find the best fit package in the family (if any)
+    MddCore::PackageId bestFit;
+    winrt::Windows::Management::Deployment::PackageManager packageManager;
+    for (const auto& candidatePackageFullName : foundPackageFullNames)
+    {
+        auto candidate{ MddCore::PackageId::FromPackageFullName(candidatePackageFullName.c_str()) };
+
+        // Do we already have a higher version under consideration?
+        if (bestFit && (bestFit.Version().Version > candidate.Version().Version))
+        {
+            continue;
+        }
+
+        // Package version must meet the minVersion filter
+        if (candidate.Version().Version < packageDependency.MinVersion().Version)
+        {
+            continue;
+        }
+
+        // Package architecture must meet the architecture filter
+        if (packageDependency.Architectures() == MddPackageDependencyProcessorArchitectures::None)
+        {
+            if (!IsPackageABetterFitPerArchitecture(bestFit, candidate))
+            {
+                continue;
+            }
+        }
+        else
+        {
+            if (!packageDependency.IsArchitectureInArchitectures(candidate.Architecture()))
+            {
+                continue;
+            }
+        }
+
+        // Package status must be OK to use a package
+        winrt::hstring currentUser;
+        auto package{ packageManager.FindPackageForUser(currentUser, candidate.PackageFullName()) };
+        auto status{ package.Status() };
+        if (!status.VerifyIsOK())
+        {
+            continue;
+        }
+
+        // The new candidate is better than the current champion
+        bestFit = std::move(candidate);
+    }
+
+    // Did we fail to find a match?
+    THROW_HR_IF(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), bestFit.PackageFullName().empty());
+
+    // We have a winner!
+    packageFullName = std::move(wil::make_process_heap_string(bestFit.PackageFullName().c_str()));
+    return S_OK;
+}
+
+HRESULT MddCore::PackageGraph::Remove(
+    MDD_PACKAGEDEPENDENCY_CONTEXT context)
+{
+    for (size_t index=0; index < m_packageGraphNodes.size(); ++index)
+    {
+        auto& node{ m_packageGraphNodes[index] };
+        if (node.Context() == context)
+        {
+            // Detach the node from the package graph before updating the DLL Search Order
+            auto detachedNode{ std::move(node) };
+            m_packageGraphNodes.erase(m_packageGraphNodes.begin() + index);
+
+            // The DLL Search Order must be updated when we update the package graph
+            RemoveFromDllSearchOrder(detachedNode);
+
+            return S_OK;
+        }
+    }
+    RETURN_WIN32(ERROR_INVALID_HANDLE);
+}
+
+HRESULT MddCore::PackageGraph::GetPackageDependencyForContext(
+    _In_ MDD_PACKAGEDEPENDENCY_CONTEXT context,
+    wil::unique_process_heap_string& packageDependencyId)
+{
+    for (size_t index=0; index < m_packageGraphNodes.size(); ++index)
```

Comment:
ranged-based for

Reply:
I had multiple problems with range-based for (and the error messages are disastrously unhelpful), though that may have been to other (reasonable) problems during coding now long-since resolved. Will take a pass at range-based for in a future PR.

Reply:
fine by me. :-)